executor/analyze.go

```diff
@@ -86,6 +89,24 @@ const (
 	maxSketchSize       = 10000
 )
 
+var (
+	memToGC = atomicutil.NewInt64(0)
+	inGC    = atomicutil.NewBool(false)
+)
+
+func tryGC(memTracker *memory.Tracker, force bool) {
+	if inGC.CAS(false, true) {
+		defer inGC.Store(false)
+		toGC := memToGC.Load()
+		gcTrigger := variable.AnalyzeGCTrigger.Load()
+		if force || (gcTrigger > 0 && toGC > gcTrigger) {
+			runtime.GC()
+			memTracker.Consume(-toGC)
```

Comment:
I think we should separate `memtracker.Consume` with gc as we recorded different object memory size, eg:

In `subMergeWorker`, we only record `colRespSize` in memtracker while we both record `colRespSize` and `dataSize` in memToGC

```golang
                 dataSize := int64(cap(data))
		colResp := &tipb.AnalyzeColumnsResp{}
		err := colResp.Unmarshal(data)
		if err != nil {
			resultCh <- &samplingMergeResult{err: err}
			return
		}
		colRespSize := int64(colResp.Size())
		memTracker.Consume(colRespSize)
		subCollector := statistics.NewRowSampleCollector(int(e.analyzePB.ColReq.SampleSize), e.analyzePB.ColReq.GetSampleRate(), l)
		tmpMemSize := subCollector.Base().FromProto(colResp.RowCollector, e.memTracker)
		UpdateAnalyzeJob(e.ctx, e.job, subCollector.Base().Count)
		oldRetCollectorSize := retCollector.Base().MemSize
		retCollector.MergeCollector(subCollector)
		newRetCollectorSize := retCollector.Base().MemSize
		subCollectorSize := subCollector.Base().MemSize
		memToGC.Add(dataSize + colRespSize + tmpMemSize)
		memToGC.Add(oldRetCollectorSize + subCollectorSize - newRetCollectorSize)
```

Reply:
dataSize is recorded when read and send data to taskCh:
```
func readDataAndSendTask(ctx sessionctx.Context, handler *tableResultHandler, mergeTaskCh chan []byte, memTracker *memory.Tracker) error {
	defer close(mergeTaskCh)
	for {
		failpoint.Inject("mockKillRunningV2AnalyzeJob", func() {
			dom := domain.GetDomain(ctx)
			dom.SysProcTracker().KillSysProcess(util.GetAutoAnalyzeProcID(dom.ServerID))
		})
		if atomic.LoadUint32(&ctx.GetSessionVars().Killed) == 1 {
			return errors.Trace(ErrQueryInterrupted)
		}
		failpoint.Inject("mockSlowAnalyzeV2", func() {
			time.Sleep(1000 * time.Second)
		})
		data, err := handler.nextRaw(context.TODO())
		if err != nil {
			return errors.Trace(err)
		}
		if data == nil {
			break
		}
		memTracker.Consume(int64(cap(data)))
		mergeTaskCh <- data
	}
	return nil
}
```